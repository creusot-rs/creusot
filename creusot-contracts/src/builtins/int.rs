use super::model::*;
use creusot_contracts_proc::*;
use std::ops::*;

#[rustc_diagnostic_item = "creusot_int"]
pub struct Int;

impl PartialEq for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "eq_int"]
    fn eq(&self, _: &Int) -> bool {
        panic!()
    }

    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "ne_int"]
    fn ne(&self, _: &Int) -> bool {
        panic!()
    }
}

impl PartialOrd for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "le_int"]
    fn le(&self, _: &Int) -> bool {
        panic!()
    }

    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "ge_int"]
    fn ge(&self, _: &Int) -> bool {
        panic!()
    }

    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "lt_int"]
    fn lt(&self, _: &Int) -> bool {
        panic!()
    }

    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "gt_int"]
    fn gt(&self, _: &Int) -> bool {
        panic!()
    }

    #[creusot::spec::no_translate]
    fn partial_cmp(&self, _: &Self) -> Option<std::cmp::Ordering> {
        panic!()
    }
}

impl From<u32> for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "u32_to_int"]
    fn from(_: u32) -> Self {
        panic!()
    }
}
impl Model for u32 {
    type ModelTy = Int;
    #[logic]
    #[rustc_diagnostic_item = "u32_model"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<i32> for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "i32_to_int"]
    fn from(_: i32) -> Self {
        panic!()
    }
}
impl Model for i32 {
    type ModelTy = Int;
    #[logic]
    #[rustc_diagnostic_item = "i32_model"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<usize> for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "usize_to_int"]
    fn from(_: usize) -> Self {
        panic!()
    }
}
impl Model for usize {
    type ModelTy = Int;
    #[logic]
    #[rustc_diagnostic_item = "usize_model"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<isize> for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "isize_to_int"]
    fn from(_: isize) -> Self {
        panic!()
    }
}
impl Model for isize {
    type ModelTy = Int;
    #[logic]
    #[rustc_diagnostic_item = "isize_model"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl Add<Int> for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "add_int"]
    fn add(self, _: Int) -> Self {
        panic!()
    }
}

impl Sub<Int> for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "sub_int"]
    fn sub(self, _: Int) -> Self {
        panic!()
    }
}

impl Mul<Int> for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "mul_int"]
    fn mul(self, _: Int) -> Self {
        panic!()
    }
}

impl Div<Int> for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "div_int"]
    fn div(self, _: Int) -> Self {
        panic!()
    }
}

impl Rem<Int> for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "rem_int"]
    fn rem(self, _: Int) -> Self {
        panic!()
    }
}

impl Neg for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "neg_int"]
    fn neg(self) -> Self {
        panic!()
    }
}
