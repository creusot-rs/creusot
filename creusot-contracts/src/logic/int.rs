use super::model::*;
use crate::logic::*;

use creusot_contracts_proc::*;

use std::ops::*;

#[rustc_diagnostic_item = "creusot_int"]
pub struct Int(*mut ());

impl WellFounded for Int {}

impl Int {
    #[trusted]
    #[logic]
    #[creusot::builtins = "int.Power.power"]
    pub fn pow(self, _: Int) -> Int {
        std::process::abort()
    }
}

impl From<u32> for Int {
    #[logic]
    #[trusted]
    #[rustc_diagnostic_item = "u32_to_int"]
    #[creusot::builtins = "mach.int.UInt32.to_int"]
    fn from(_: u32) -> Self {
        std::process::abort()
    }
}
impl Model for u32 {
    type ModelTy = Int;
    #[logic]
    #[rustc_diagnostic_item = "u32_model"]
    #[creusot::builtins = "mach.int.UInt32.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<i32> for Int {
    #[logic]
    #[trusted]
    #[rustc_diagnostic_item = "i32_to_int"]
    #[creusot::builtins = "mach.int.Int32.to_int"]
    fn from(_: i32) -> Self {
        std::process::abort()
    }
}
impl Model for i32 {
    type ModelTy = Int;
    #[logic]
    #[rustc_diagnostic_item = "i32_model"]
    #[creusot::builtins = "mach.int.Int32.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<usize> for Int {
    #[logic]
    #[trusted]
    #[rustc_diagnostic_item = "usize_to_int"]
    #[creusot::builtins = "mach.int.UInt64.to_int"]
    fn from(_: usize) -> Self {
        std::process::abort()
    }
}
impl Model for usize {
    type ModelTy = Int;
    #[logic]
    #[rustc_diagnostic_item = "usize_model"]
    #[creusot::builtins = "mach.int.UInt64.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<isize> for Int {
    #[logic]
    #[trusted]
    #[rustc_diagnostic_item = "isize_to_int"]
    fn from(_: isize) -> Self {
        std::process::abort()
    }
}
impl Model for isize {
    type ModelTy = Int;
    #[logic]
    #[rustc_diagnostic_item = "isize_model"]
    #[creusot::builtins = "mach.int.Int64.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl Add<Int> for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[rustc_diagnostic_item = "add_int"]
    fn add(self, _: Int) -> Self {
        panic!()
    }
}

impl Sub<Int> for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[rustc_diagnostic_item = "sub_int"]
    fn sub(self, _: Int) -> Self {
        panic!()
    }
}

impl Mul<Int> for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[rustc_diagnostic_item = "mul_int"]
    fn mul(self, _: Int) -> Self {
        panic!()
    }
}

impl Div<Int> for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[rustc_diagnostic_item = "div_int"]
    fn div(self, _: Int) -> Self {
        panic!()
    }
}

impl Rem<Int> for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[rustc_diagnostic_item = "rem_int"]
    fn rem(self, _: Int) -> Self {
        panic!()
    }
}

impl Neg for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[rustc_diagnostic_item = "neg_int"]
    fn neg(self) -> Self {
        panic!()
    }
}
