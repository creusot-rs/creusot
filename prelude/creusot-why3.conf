[prover_modifiers]
name = "Z3"
driver = "z3_nombqi_seq.drv"