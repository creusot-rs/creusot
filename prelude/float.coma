module Float32
  use int.Int
  use bv.Pow2int
  use real.Abs as Abs
  use real.FromInt as FromInt
  use real.Truncate as Truncate
  use real.RealInfix
  use export ieee_float.RoundingMode

  type t = < float 8 24 >

  (** {2 Part I - Public Interface}   *)

  constant eb : int = t'eb
  (** the number of bits in the exponent.  *)

  constant sb : int = t'sb
  (** the number of bits in the significand, including the hidden bit. *)

  axiom eb_gt_1: 1 < eb
  axiom sb_gt_1: 1 < sb

  (** {3 Constructors and Constants} *)

  constant zeroF     : t   (** +0.0 *)
  (* exp_bias = 2^(eb - 1) - 1 *)
  (* max_finite_exp = 2^sb - 2 - exp_bias = exp_bias *)
  (* max_significand = (2^eb + 2^eb - 1) * 2^(1-eb) *)
  (* max_value = (2^eb + 2^eb - 1) * 2^(1-eb) * 2 ^ max_finite_exp *)
  (* [m;x] = ( 1 + m * 2^(1-eb) ) * 2^( x - exp_bias ) *)
  (* max_value = [2^(eb-1); 2^sb - 2] *)

  (** {3 Operators} *)
  let addi (mode : mode) (a : t) (b : t) (ret (result : t)) = any
  function addi mode t t : t
  let subi (mode : mode) (a : t) (b : t) (ret (result : t)) = any
  function subi mode t t : t
  let muli (mode : mode) (a : t) (b : t) (ret (result : t)) = any
  function muli mode t t : t
  let divi (mode : mode) (a : t) (b : t) (ret (result : t)) = any
  function divi mode t t : t
    (** The four basic operations, rounded in the given mode *)


  let abs (a : t) (ret (result : t)) = any
  function abs         t   : t   (** Absolute value *)

  let fma  (mode : mode) (a : t) (b : t) (c : t) (ret (result : t)) = any
  function fma  mode t t t : t   (** Fused multiply-add: x * y + z *)
  let sqrt (mode : mode) (a : t) (ret (result : t)) = any
  function sqrt mode   t   : t   (** Square root *)

  function neg (x : t) : t
  let neg (x : t) (ret (result : t)) = ret {neg x}

  function add (x : t) (y : t) : t = addi RNE x y
  let add (a : t) (b : t) (ret (result : t)) = ret {add a b}

  function sub (x : t) (y : t) : t = subi RNE x y
  let sub (a : t) (b : t) (ret (result : t)) = ret {sub a b}

  function mul (x : t) (y : t) : t = muli RNE x y
  let mul (a : t) (b : t) (ret (result : t)) = ret {mul a b}

  function div (x : t) (y : t) : t = divi RNE x y
  let div (a : t) (b : t) (ret (result : t)) = ret {div a b}

    (** Notations for operations in the default mode RNE *)

  function roundToIntegral mode t : t
  let roundToIntegral (mode : mode) (a : t) (ret (result : t) {result = roundToIntegral mode a}) = any
    (** Rounding to an integer *)

  function min       t t : t
  function max       t t : t

  predicate le t t
  predicate lt t t
  predicate eq t t
  predicate ge (x:t) (y:t) = le y x
  predicate gt (x:t) (y:t) = lt y x

  let le (a : t) (b : t) (ret (result : bool) { result = le a b }) = any
  let lt (a : t) (b : t) (ret (result : bool) { result = lt a b }) = any
  let eq (a : t) (b : t) (ret (result : bool) { result = eq a b }) = any
  let ge (a : t) (b : t) (ret (result : bool) { result = ge a b }) = any
  let gt (a : t) (b : t) (ret (result : bool) { result = gt a b }) = any
end
module Float64
  use int.Int
  use bv.Pow2int
  use real.Abs as Abs
  use real.FromInt as FromInt
  use real.Truncate as Truncate
  use real.RealInfix
  use export ieee_float.RoundingMode

  type t = < float 11 53 >

  (** {2 Part I - Public Interface}   *)

  constant eb : int = t'eb
  (** the number of bits in the exponent.  *)

  constant sb : int = t'sb
  (** the number of bits in the significand, including the hidden bit. *)

  axiom eb_gt_1: 1 < eb
  axiom sb_gt_1: 1 < sb

  (** {3 Constructors and Constants} *)

  constant zeroF     : t   (** +0.0 *)
  (* exp_bias = 2^(eb - 1) - 1 *)
  (* max_finite_exp = 2^sb - 2 - exp_bias = exp_bias *)
  (* max_significand = (2^eb + 2^eb - 1) * 2^(1-eb) *)
  (* max_value = (2^eb + 2^eb - 1) * 2^(1-eb) * 2 ^ max_finite_exp *)
  (* [m;x] = ( 1 + m * 2^(1-eb) ) * 2^( x - exp_bias ) *)
  (* max_value = [2^(eb-1); 2^sb - 2] *)

  (** {3 Operators} *)
  let addi (mode : mode) (a : t) (b : t) (ret (result : t)) = any
  function addi mode t t : t
  let subi (mode : mode) (a : t) (b : t) (ret (result : t)) = any
  function subi mode t t : t
  let muli (mode : mode) (a : t) (b : t) (ret (result : t)) = any
  function muli mode t t : t
  let divi (mode : mode) (a : t) (b : t) (ret (result : t)) = any
  function divi mode t t : t
    (** The four basic operations, rounded in the given mode *)


  let abs (a : t) (ret (result : t)) = any
  function abs         t   : t   (** Absolute value *)

  let fma  (mode : mode) (a : t) (b : t) (c : t) (ret (result : t)) = any
  function fma  mode t t t : t   (** Fused multiply-add: x * y + z *)
  let sqrt (mode : mode) (a : t) (ret (result : t)) = any
  function sqrt mode   t   : t   (** Square root *)

  function neg (x : t) : t
  let neg (x : t) (ret (result : t)) = ret {neg x}

  function add (x : t) (y : t) : t = addi RNE x y
  let add (a : t) (b : t) (ret (result : t)) = ret {add a b}

  function sub (x : t) (y : t) : t = subi RNE x y
  let sub (a : t) (b : t) (ret (result : t)) = ret {sub a b}

  function mul (x : t) (y : t) : t = muli RNE x y
  let mul (a : t) (b : t) (ret (result : t)) = ret {mul a b}

  function div (x : t) (y : t) : t = divi RNE x y
  let div (a : t) (b : t) (ret (result : t)) = ret {div a b}

    (** Notations for operations in the default mode RNE *)

  function roundToIntegral mode t : t
  let roundToIntegral (mode : mode) (a : t) (ret (result : t) {result = roundToIntegral mode a}) = any
    (** Rounding to an integer *)

  function min       t t : t
  function max       t t : t

  predicate le t t
  predicate lt t t
  predicate eq t t
  predicate ge (x:t) (y:t) = le y x
  predicate gt (x:t) (y:t) = lt y x

  let le (a : t) (b : t) (ret (result : bool) { result = le a b }) = any
  let lt (a : t) (b : t) (ret (result : bool) { result = lt a b }) = any
  let eq (a : t) (b : t) (ret (result : bool) { result = eq a b }) = any
  let ge (a : t) (b : t) (ret (result : bool) { result = ge a b }) = any
  let gt (a : t) (b : t) (ret (result : bool) { result = gt a b }) = any
end