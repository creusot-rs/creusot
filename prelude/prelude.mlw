module Prelude
  use Ref
  use mach.int.Int
  use mach.int.Int32
  use mach.int.Int64
  use mach.int.UInt32
  use mach.int.UInt64
  use string.Char
  use floating_point.Single
  use floating_point.Double

  type usize = uint64
  type isize = int64
  type opaque_ptr

  type borrowed 'a = { current : 'a ; final : 'a; }
  let function ( *_ ) x = x.current
  let function ( ^_ ) x = x.final
  val borrow_mut (a : 'a) : borrowed 'a
  ensures { *result = a }
end
