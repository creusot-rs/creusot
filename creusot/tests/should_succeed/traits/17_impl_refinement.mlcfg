
module C17ImplRefinement_Impl0_MyFunction_Interface
  use mach.int.UInt64
  use mach.int.Int
  use prelude.Borrow
  use prelude.UIntSize
  val my_function [@cfg:stackify] [#"../17_impl_refinement.rs" 14 4 14 34] (self : ()) : usize
    requires {[#"../17_impl_refinement.rs" 12 15 12 19] true}
    ensures { [#"../17_impl_refinement.rs" 13 14 13 27] UInt64.to_int result >= 15 }
    
end
module C17ImplRefinement_Impl0_MyFunction
  use mach.int.Int
  use prelude.UIntSize
  use prelude.Borrow
  use mach.int.UInt64
  let rec cfg my_function [@cfg:stackify] [#"../17_impl_refinement.rs" 14 4 14 34] (self : ()) : usize
    requires {[#"../17_impl_refinement.rs" 12 15 12 19] true}
    ensures { [#"../17_impl_refinement.rs" 13 14 13 27] UInt64.to_int result >= 15 }
    
   = [@vc:do_not_keep_trace] [@vc:sp]
  var _0 : usize;
  var self_1 : ();
  {
    self_1 <- self;
    goto BB0
  }
  BB0 {
    _0 <- ([#"../17_impl_refinement.rs" 15 8 15 10] (20 : usize));
    return _0
  }
  
end
module C17ImplRefinement_Impl1_NeedFalse_Stub
  use mach.int.UInt64
  use mach.int.Int
  function need_false [#"../17_impl_refinement.rs" 29 4 29 25] (y : uint64) : ()
end
module C17ImplRefinement_Impl1_NeedFalse_Interface
  use mach.int.UInt64
  use mach.int.Int
  function need_false [#"../17_impl_refinement.rs" 29 4 29 25] (y : uint64) : ()
  axiom need_false_spec : forall y : uint64 . ([#"../17_impl_refinement.rs" 28 15 28 23] UInt64.to_int y >= 15) -> true
end
module C17ImplRefinement_Impl1_NeedFalse
  use mach.int.UInt64
  use mach.int.Int
  function need_false [#"../17_impl_refinement.rs" 29 4 29 25] (y : uint64) : () =
    [#"../17_impl_refinement.rs" 27 4 27 12] ()
  val need_false [#"../17_impl_refinement.rs" 29 4 29 25] (y : uint64) : ()
    requires {[#"../17_impl_refinement.rs" 28 15 28 23] UInt64.to_int y >= 15}
    ensures { result = need_false y }
    
  axiom need_false_spec : forall y : uint64 . ([#"../17_impl_refinement.rs" 28 15 28 23] UInt64.to_int y >= 15) -> true
end
module C17ImplRefinement_Impl1_NeedFalse_Impl
  use mach.int.UInt64
  use mach.int.Int
  let rec ghost function need_false [#"../17_impl_refinement.rs" 29 4 29 25] (y : uint64) : ()
    requires {[#"../17_impl_refinement.rs" 28 15 28 23] UInt64.to_int y >= 15}
    
   = [@vc:do_not_keep_trace] [@vc:sp]
    [#"../17_impl_refinement.rs" 27 4 27 12] ()
end
module C17ImplRefinement_Tr_MyFunction_Interface
  type self
  use mach.int.UInt64
  use mach.int.Int
  use prelude.Borrow
  use prelude.UIntSize
  val my_function [@cfg:stackify] [#"../17_impl_refinement.rs" 8 4 8 35] (self : self) : usize
    ensures { [#"../17_impl_refinement.rs" 7 14 7 27] UInt64.to_int result >= 10 }
    
end
module C17ImplRefinement_Tr_MyFunction
  type self
  use mach.int.UInt64
  use mach.int.Int
  use prelude.Borrow
  use prelude.UIntSize
  val my_function [@cfg:stackify] [#"../17_impl_refinement.rs" 8 4 8 35] (self : self) : usize
    ensures { [#"../17_impl_refinement.rs" 7 14 7 27] UInt64.to_int result >= 10 }
    
end
module C17ImplRefinement_Impl0
  clone C17ImplRefinement_Impl0_MyFunction_Interface as MyFunction0
  clone C17ImplRefinement_Tr_MyFunction_Interface as MyFunction1 with
    type self = (),
    val my_function = MyFunction0.my_function
end
module C17ImplRefinement_ReqFalse_NeedFalse_Stub
  type self
  use mach.int.UInt64
  use mach.int.Int
  function need_false [#"../17_impl_refinement.rs" 22 4 22 32] (x : uint64) : ()
end
module C17ImplRefinement_ReqFalse_NeedFalse_Interface
  type self
  use mach.int.UInt64
  use mach.int.Int
  function need_false [#"../17_impl_refinement.rs" 22 4 22 32] (x : uint64) : ()
  axiom need_false_spec : forall x : uint64 . ([#"../17_impl_refinement.rs" 21 15 21 23] UInt64.to_int x >= 10) -> true
end
module C17ImplRefinement_ReqFalse_NeedFalse
  type self
  use mach.int.UInt64
  use mach.int.Int
  function need_false [#"../17_impl_refinement.rs" 22 4 22 32] (x : uint64) : ()
  val need_false [#"../17_impl_refinement.rs" 22 4 22 32] (x : uint64) : ()
    requires {[#"../17_impl_refinement.rs" 21 15 21 23] UInt64.to_int x >= 10}
    ensures { result = need_false x }
    
  axiom need_false_spec : forall x : uint64 . ([#"../17_impl_refinement.rs" 21 15 21 23] UInt64.to_int x >= 10) -> true
end
module C17ImplRefinement_Impl1
  use mach.int.UInt64
  use mach.int.Int
  clone C17ImplRefinement_Impl1_NeedFalse as NeedFalse0 with
    axiom .
  clone C17ImplRefinement_ReqFalse_NeedFalse_Interface as NeedFalse1 with
    type self = (),
    function need_false = NeedFalse0.need_false,
    axiom .
  goal need_false_spec : forall x : uint64 . (([#"../17_impl_refinement.rs" 21 15 21 23] UInt64.to_int x >= 10) -> ([#"../17_impl_refinement.rs" 28 15 28 23] UInt64.to_int x >= 15)) /\ (forall result : () . true)
end
