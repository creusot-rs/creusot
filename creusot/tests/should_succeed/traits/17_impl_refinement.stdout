module Type
  use Ref
  use mach.int.Int
  use prelude.Int8
  use prelude.Int16
  use mach.int.Int32
  use mach.int.Int64
  use prelude.UInt8
  use prelude.UInt16
  use mach.int.UInt32
  use mach.int.UInt64
  use string.Char
  use floating_point.Single
  use floating_point.Double
  use prelude.Prelude
end
module C17ImplRefinement_Tr_MyFunction_Interface
  type self   
  use mach.int.UInt64
  use mach.int.Int
  use mach.int.Int32
  use prelude.Prelude
  val my_function (self : self) : usize
    ensures { UInt64.to_int result >= 10 }
    
end
module C17ImplRefinement_Tr_MyFunction
  type self   
  use mach.int.UInt64
  use mach.int.Int
  use mach.int.Int32
  use prelude.Prelude
  val my_function (self : self) : usize
    ensures { UInt64.to_int result >= 10 }
    
end
module CreusotContracts_Logic_Resolve_Resolve_Resolve_Interface
  type self   
  predicate resolve (self : self)
end
module CreusotContracts_Logic_Resolve_Resolve_Resolve
  type self   
  predicate resolve (self : self)
end
module C17ImplRefinement_Impl0_MyFunction_Interface
  use mach.int.UInt64
  use mach.int.Int
  use mach.int.Int32
  use prelude.Prelude
  val my_function (self : ()) : usize
    requires {true}
    ensures { UInt64.to_int result >= 15 }
    
end
module C17ImplRefinement_Impl0_MyFunction
  use mach.int.UInt64
  use mach.int.Int
  use mach.int.Int32
  use prelude.Prelude
  clone CreusotContracts_Logic_Resolve_Resolve_Resolve as Resolve0 with type self = ()
  let rec cfg my_function (self : ()) : usize
    requires {true}
    ensures { UInt64.to_int result >= 15 }
    
   = 
  var _0 : usize;
  var self_1 : ();
  {
    self_1 <- self;
    goto BB0
  }
  BB0 {
    assume { Resolve0.resolve self_1 };
    _0 <- (20 : usize);
    return _0
  }
  
end
module C17ImplRefinement_Impl0
  clone C17ImplRefinement_Impl0_MyFunction_Interface as MyFunction0
  clone C17ImplRefinement_Tr_MyFunction_Interface as MyFunction1 with type self = (),
  val my_function = MyFunction0.my_function
end
